Null subject in Query can cause SQLiteException

```
java.lang.RuntimeException: An error occured while executing doInBackground()
            at android.os.AsyncTask$3.done(AsyncTask.java:299)
            at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:352)
            at java.util.concurrent.FutureTask.setException(FutureTask.java:219)
            at java.util.concurrent.FutureTask.run(FutureTask.java:239)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573)
            at java.lang.Thread.run(Thread.java:841)
     Caused by: android.database.sqlite.SQLiteException: near "ORDER": syntax error (code 1): , while compiling: SELECT objects.rowid AS `_id`, objects.bucket || objects.key AS `key`, objects.key as `object_key`, objects.data as `object_data` , i2.value AS `title`, i3.value AS `contentPreview`, i4.value AS `pinned` FROM `objects`  LEFT JOIN indexes AS i0 ON objects.bucket = i0.bucket AND objects.key = i0.key AND i0.name=? LEFT JOIN indexes AS i1 ON objects.bucket = i1.bucket AND objects.key = i1.key AND i1.name=? LEFT JOIN indexes AS i2 ON objects.bucket = i2.bucket AND objects.key = i2.key AND i2.name=? LEFT JOIN indexes AS i3 ON objects.bucket = i3.bucket AND objects.key = i3.key AND i3.name=? LEFT JOIN indexes AS i4 ON objects.bucket = i4.bucket AND objects.key = i4.key AND i4.name=? LEFT JOIN indexes AS i5 ON objects.bucket = i5.bucket AND objects.key = i5.key AND i5.name=? WHERE objects.bucket = ? AND (  i0.value IS NULL OR i0.value !=  1) AND (  i1.value IS NOT NULL AND i1.value LIKE  ORDER BY i4.value DESC,  i5.value DESC
            at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
            at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:889)
            at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:500)
            at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
            at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:58)
            at android.database.sqlite.SQLiteQuery.<init>(SQLiteQuery.java:37)
            at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:44)
            at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1314)
            at android.database.sqlite.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1253)
            at com.simperium.android.PersistentStore$QueryBuilder.query(PersistentStore.java:433)
            at com.simperium.android.PersistentStore$DataStore.search(PersistentStore.java:161)
            at com.simperium.client.Bucket.searchObjects(Bucket.java:345)
            at com.simperium.client.Query.execute(Query.java:284)
            at com.automattic.simplenote.NoteListFragment.queryNotes(NoteListFragment.java:443)
            at com.automattic.simplenote.NoteListFragment$refreshListTask.doInBackground(NoteListFragment.java:666)
            at com.automattic.simplenote.NoteListFragment$refreshListTask.doInBackground(NoteListFragment.java:660)
            at android.os.AsyncTask$2.call(AsyncTask.java:287)
            at java.util.concurrent.FutureTask.run(FutureTask.java:234)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573)
            at java.lang.Thread.run(Thread.java:841)
```

--- a/Simperium/src/instrumentTest/java/com/simperium/LoginActivityTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/LoginActivityTest.java
@@ -36,6 +36,7 @@ public class LoginActivityTest extends ActivityInstrumentationTestCase2<LoginAct
     @Override
     protected void setUp() throws Exception {
         super.setUp();
+        setActivityInitialTouchMode(false);
         mClient = new MockClient();
         mSimperium = Simperium.newClient("fake-app", "fake-secret", mClient);
     }

--- a/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
@@ -259,6 +259,19 @@ public class PersistentStoreTest extends ActivityInstrumentationTestCase2<LoginA
         assertEquals(10, cursor.getCount());
     }
 
+    public void testQueryWithNullSubject()
+    throws Exception {
+
+        mActivity.deleteDatabase(mDatabaseName);
+        Note.Schema schema = new Note.Schema();
+
+        Query<Note> query = new Query<Note>(mBucket);
+
+        Bucket.ObjectCursor<Note> cursor = query.where("title", Query.ComparisonType.LIKE, null).execute();
+
+        cursor.close();
+    }
+
     public void testCounts()
     throws Exception {
         // clear out the database

--- a/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
@@ -234,15 +234,16 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.close();
     }
 
+    /**
+     * Search for notes with a null value in "null_column" index
+     */
     public void testSearchForNull() {
-
-        // clear out the database
-        mActivity.deleteDatabase(mDatabaseName);
 
         Note note = mBucket.newObject();
         note.setTitle("Special will be null");
+        note.save();
 
-        int count = mBucket.query().where("special", Query.ComparisonType.EQUAL_TO, null).count();
+        int count = mBucket.query().where("null_column", Query.ComparisonType.EQUAL_TO, null).count();
 
         assertEquals(1, count);
 

--- a/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
@@ -141,7 +141,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         note.save();
 
         Cursor cursor = mDatabase.query(PersistentStore.INDEXES_TABLE, null, null, null, null, null, "name", null);
-        assertEquals(8, cursor.getCount());
+        assertEquals(9, cursor.getCount());
         cursor.moveToFirst();
         assertEquals(bucketName, cursor.getString(0));
         assertEquals(note.getSimperiumKey(), cursor.getString(1));

--- a/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
@@ -1,4 +1,4 @@
-package com.simperium;
+package com.simperium.android;
 
 import android.database.Cursor;
 import android.database.sqlite.SQLiteDatabase;
@@ -28,55 +28,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 
-import static com.simperium.TestHelpers.makeUser;
-
-public class PersistentStoreTest extends ActivityInstrumentationTestCase2<LoginActivity> {
-    public static final String TAG = "SimperiumTest.PersistentStore";
-    public static final String MASTER_TABLE = "sqlite_master";
-    public static final String BUCKET_NAME="bucket";
-    
-    private LoginActivity mActivity;
-    
-    private PersistentStore mStore;
-    private BucketStore<Note> mNoteStore;
-    private SQLiteDatabase mDatabase;
-    private String mDatabaseName = "simperium-test-data";
-    private String[] mTableNames = new String[]{"indexes", "objects", "value_caches"};
-    private Bucket<Note> mBucket;
-    private User mUser;
-    private BucketSchema mSchema;
-    private ObjectCache<Note> mCache;
-    private GhostStorageProvider mGhostStore;
-
-    public PersistentStoreTest() {
-        super(LoginActivity.class);
-    }
-
-    @Override
-    protected void setUp() throws Exception {
-
-        super.setUp();
-
-        setActivityInitialTouchMode(false);
-        mUser = makeUser();
-        mActivity = getActivity();
-        mDatabase = mActivity.openOrCreateDatabase(mDatabaseName, 0, null);
-        mGhostStore = new MockGhostStore();
-        mCache = new MockCache<Note>();
-        mStore = new PersistentStore(mDatabase);
-        mSchema = new Note.Schema();
-        mNoteStore = mStore.createStore(BUCKET_NAME, mSchema);
-        mBucket = new Bucket<Note>(MockExecutor.immediate(), BUCKET_NAME, mSchema, mUser, mNoteStore, mGhostStore, mCache);
-        Bucket.Channel channel = new MockChannel(mBucket);
-        mBucket.setChannel(channel);
-        mNoteStore.prepare(mBucket);
-    }
-
-    @Override
-    protected void tearDown() throws Exception {
-        mActivity.deleteDatabase(mDatabaseName);
-        super.tearDown();
-    }
+public class PersistentStoreTest extends PersistentStoreBaseTest {
 
     public void testDatabaseTables()
     throws Exception {
@@ -268,6 +220,20 @@ public class PersistentStoreTest extends ActivityInstrumentationTestCase2<LoginA
         Bucket.ObjectCursor<Note> cursor = query.where("title", Query.ComparisonType.LIKE, null).execute();
 
         cursor.close();
+    }
+
+    public void testSearchForNull() {
+
+        // clear out the database
+        mActivity.deleteDatabase(mDatabaseName);
+
+        Note note = mBucket.newObject();
+        note.setTitle("Special will be null");
+
+        int count = mBucket.query().where("special", Query.ComparisonType.EQUAL_TO, null).count();
+
+        assertEquals(1, count);
+
     }
 
     public void testCounts()

--- a/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
@@ -53,6 +53,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         assertEquals(bucketName, cursor.getString(1));
         assertEquals(key, cursor.getString(2));
         assertEquals("{\"tags\":[],\"deleted\":false,\"title\":\"Hola Mundo!\"}", cursor.getString(3));
+        cursor.close();
     }
   
     public void testDeletingObject()
@@ -67,6 +68,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
   
         Cursor cursor = mStore.queryObject(bucketName, key);
         assertEquals(0, cursor.getCount());
+        cursor.close();
     }
   
     public void testGettingObject()
@@ -100,6 +102,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         
         Bucket.ObjectCursor<Note> cursor = mNoteStore.all();
         assertEquals(2, cursor.getCount());
+        cursor.close();
     }
   
     public void testResetData()
@@ -112,12 +115,13 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
   
         Bucket.ObjectCursor<Note> cursor = mNoteStore.all();
         assertEquals(1, cursor.getCount());
+        cursor.close();
   
         mNoteStore.reset();
   
         cursor = mNoteStore.all();
         assertEquals(0, cursor.getCount());
-  
+        cursor.close();
     }
   
     public void testIndexObject()
@@ -157,6 +161,8 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.moveToNext();
         assertEquals("col4", cursor.getString(2));
         assertEquals("dos", cursor.getString(3));
+
+        cursor.close();
     }
 
     public void testObjectSearching()
@@ -183,32 +189,38 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         query.where("special", Query.ComparisonType.EQUAL_TO, true);
         cursor = store.search(query);
         assertEquals(20, cursor.getCount());
+        cursor.close();
 
         query = new Query<Note>();
         query.where("title", Query.ComparisonType.LIKE, "Note 7%");
         cursor = store.search(query);
         assertEquals(111, cursor.getCount());
-        
+        cursor.close();
+
         query = new Query<Note>();
         query.where("title", Query.ComparisonType.NOT_LIKE, "Note 7%");
         cursor = store.search(query);
         assertEquals(889, cursor.getCount());
-        
+        cursor.close();
+
         query = new Query<Note>();
         query.where("special", Query.ComparisonType.NOT_EQUAL_TO, true);
         cursor = store.search(query);
         assertEquals(980, cursor.getCount());
-        
+        cursor.close();
+
         query = new Query<Note>();
         query.where("special", Query.ComparisonType.NOT_EQUAL_TO, false);
         query.where("title", Query.ComparisonType.LIKE, "Note 1%");
         cursor = store.search(query);
         assertEquals(110, cursor.getCount());
+        cursor.close();
 
         query = new Query<Note>();
         query.where("spanish", Query.ComparisonType.EQUAL_TO, "cuatro");
         cursor = store.search(query);
         assertEquals(10, cursor.getCount());
+        cursor.close();
     }
 
     public void testQueryWithNullSubject()
@@ -298,7 +310,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.moveToFirst();
         Note note = cursor.getObject();
         assertEquals(1, note.get("position"));
-        
+        cursor.close();
 
         query = new Query<Note>();
         query.order("title", Query.SortType.DESCENDING);
@@ -311,6 +323,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.moveToNext();
         note = cursor.getObject();
         assertEquals(2, note.get("position"));
+        cursor.close();
 
         query = new Query<Note>();
         query.order("position");
@@ -320,6 +333,8 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.moveToFirst();
         note = cursor.getObject();
         assertEquals(2, note.get("position"));
+        cursor.close();
+
     }
 
     /**
@@ -336,6 +351,7 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.moveToFirst();
         assertEquals(5, cursor.getColumnCount());
         assertEquals("Lol", cursor.getString(4));
+        cursor.close();
     }
 
     public void testFullTextSearching()
@@ -376,12 +392,13 @@ public class PersistentStoreTest extends PersistentStoreBaseTest {
         cursor.moveToFirst();
 
         assertEquals("Hello <match>world</match>. Hola mundo. The <match>world</match> is your oyster. Lorem ipsum dolor sit amet, consectetur\u2026", cursor.getString(cursor.getColumnIndexOrThrow("match")));
-
+        cursor.close();
     }
 
     public static void assertTableExists(SQLiteDatabase database, String tableName){
         Cursor cursor = database.query(MASTER_TABLE, new String[]{"name"}, "type=? AND name=?", new String[]{"table", tableName}, "name", null, null, null);
         assertEquals(String.format("Table %s does not exist in %s", tableName, database), 1, cursor.getCount());
+        cursor.close();
     }
 
     private class DatabaseHelper extends SQLiteOpenHelper {

--- a/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/PersistentStoreTest.java
@@ -262,11 +262,9 @@ public class PersistentStoreTest extends ActivityInstrumentationTestCase2<LoginA
     public void testQueryWithNullSubject()
     throws Exception {
 
-        mActivity.deleteDatabase(mDatabaseName);
-        Note.Schema schema = new Note.Schema();
-
         Query<Note> query = new Query<Note>(mBucket);
 
+        // with issue #90 this would throw android.database.sqlite.SQLiteException
         Bucket.ObjectCursor<Note> cursor = query.where("title", Query.ComparisonType.LIKE, null).execute();
 
         cursor.close();

--- a/Simperium/src/instrumentTest/java/com/simperium/android/PersistentStoreBaseTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/android/PersistentStoreBaseTest.java
@@ -1,0 +1,72 @@
+package com.simperium.android;
+
+import android.database.sqlite.SQLiteDatabase;
+import android.test.ActivityInstrumentationTestCase2;
+
+import com.simperium.client.Bucket;
+import com.simperium.client.BucketSchema;
+import com.simperium.client.GhostStorageProvider;
+import com.simperium.client.ObjectCacheProvider.ObjectCache;
+import com.simperium.client.Query;
+import com.simperium.client.User;
+
+import com.simperium.storage.StorageProvider.BucketStore;
+import com.simperium.models.Note;
+
+import com.simperium.test.MockCache;
+import com.simperium.test.MockChannel;
+import com.simperium.test.MockGhostStore;
+import com.simperium.test.MockExecutor;
+
+import static com.simperium.TestHelpers.makeUser;
+
+public abstract class PersistentStoreBaseTest extends ActivityInstrumentationTestCase2<LoginActivity> {
+
+    public static final String MASTER_TABLE = "sqlite_master";
+    public static final String BUCKET_NAME="bucket";
+
+    protected LoginActivity mActivity;
+    
+    protected PersistentStore mStore;
+    protected BucketStore<Note> mNoteStore;
+    protected SQLiteDatabase mDatabase;
+    protected String mDatabaseName = "simperium-test-data";
+    protected String[] mTableNames = new String[]{"indexes", "objects", "value_caches"};
+    protected Bucket<Note> mBucket;
+    protected User mUser;
+    protected BucketSchema mSchema;
+    protected ObjectCache<Note> mCache;
+    protected GhostStorageProvider mGhostStore;
+
+    public PersistentStoreBaseTest() {
+        super(LoginActivity.class);
+    }
+
+    @Override
+    protected void setUp()
+    throws Exception {
+
+        super.setUp();
+
+        setActivityInitialTouchMode(false);
+        mUser = makeUser();
+        mActivity = getActivity();
+        mDatabase = mActivity.openOrCreateDatabase(mDatabaseName, 0, null);
+        mGhostStore = new MockGhostStore();
+        mCache = new MockCache<Note>();
+        mStore = new PersistentStore(mDatabase);
+        mSchema = new Note.Schema();
+        mNoteStore = mStore.createStore(BUCKET_NAME, mSchema);
+        mBucket = new Bucket<Note>(MockExecutor.immediate(), BUCKET_NAME, mSchema, mUser, mNoteStore, mGhostStore, mCache);
+        Bucket.Channel channel = new MockChannel(mBucket);
+        mBucket.setChannel(channel);
+        mNoteStore.prepare(mBucket);
+    }
+
+    @Override
+    protected void tearDown() throws Exception {
+        mActivity.deleteDatabase(mDatabaseName);
+        super.tearDown();
+    }
+
+}

--- a/Simperium/src/instrumentTest/java/com/simperium/android/PersistentStoreQueryBuilderTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/android/PersistentStoreQueryBuilderTest.java
@@ -1,0 +1,54 @@
+package com.simperium.android;
+
+import android.test.MoreAsserts;
+
+import com.simperium.android.PersistentStore.QueryBuilder;
+import com.simperium.android.PersistentStore.DataStore;
+import com.simperium.client.Query;
+
+import com.simperium.models.Note;
+
+public class PersistentStoreQueryBuilderTest extends PersistentStoreBaseTest {
+
+    public void testEqualToComparisonQuery()
+    throws Exception {
+
+        Query query = new Query();
+        query.where("title", Query.ComparisonType.EQUAL_TO, "lol");
+        QueryBuilder builder = buildQuery(query);
+
+        String[] args = new String[] {
+            "title", // index name
+            "bucket", // bucket name
+            "lol" // value to compare to
+        };
+        String condition = " FROM `objects`  LEFT JOIN indexes AS i0 ON objects.bucket = i0.bucket AND objects.key = i0.key AND i0.name=? WHERE objects.bucket = ? AND (  i0.value IS NOT NULL AND i0.value =  ?) ";
+
+        MoreAsserts.assertEquals(args, builder.args);
+        assertEquals(condition.toString(), builder.statement.toString());
+
+
+    }
+
+    public void testEqualToNullComparisonQuery() {
+
+        Query query = new Query();
+        query.where("title", Query.ComparisonType.EQUAL_TO, null);
+        QueryBuilder builder = buildQuery(query);
+
+        String[] args = new String[] {
+            "title", // index name
+            "bucket", // bucket name
+        };
+        String condition = " FROM `objects`  LEFT JOIN indexes AS i0 ON objects.bucket = i0.bucket AND objects.key = i0.key AND i0.name=? WHERE objects.bucket = ? AND (  i0.value IS NULL ) ";
+
+        MoreAsserts.assertEquals(args, builder.args);
+        assertEquals(condition.toString(), builder.statement.toString());
+
+    }
+
+    protected QueryBuilder buildQuery(Query query) {
+        return new QueryBuilder((DataStore<Note>) mNoteStore, query);
+    }
+
+}

--- a/Simperium/src/instrumentTest/java/com/simperium/android/PersistentStoreQueryBuilderTest.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/android/PersistentStoreQueryBuilderTest.java
@@ -40,7 +40,7 @@ public class PersistentStoreQueryBuilderTest extends PersistentStoreBaseTest {
             "title", // index name
             "bucket", // bucket name
         };
-        String condition = " FROM `objects`  LEFT JOIN indexes AS i0 ON objects.bucket = i0.bucket AND objects.key = i0.key AND i0.name=? WHERE objects.bucket = ? AND (  i0.value IS NULL ) ";
+        String condition = " FROM `objects`  LEFT JOIN indexes AS i0 ON objects.bucket = i0.bucket AND objects.key = i0.key AND i0.name=? WHERE objects.bucket = ? AND ( i0.value IS NULL )  ";
 
         MoreAsserts.assertEquals(args, builder.args);
         assertEquals(condition.toString(), builder.statement.toString());

--- a/Simperium/src/instrumentTest/java/com/simperium/models/Note.java
+++ b/Simperium/src/instrumentTest/java/com/simperium/models/Note.java
@@ -18,6 +18,7 @@ public class Note extends BucketObject {
         public Schema(){
             autoIndex();
             addIndex(contentIndexer);
+            addIndex(nullIndexer);
             setupFullTextIndex("tags", "content");
             setDefault("tags", new JSONArray());
             setDefault("deleted", false);
@@ -28,6 +29,15 @@ public class Note extends BucketObject {
             public List<Index> index(Note note){
                 List<Index> indexes = new ArrayList<Index>(1);
                 indexes.add(new Index("preview", note.getPreview()));
+                return indexes;
+            }
+        };
+
+        private Indexer<Note> nullIndexer = new Indexer<Note>() {
+            @Override
+            public List<Index> index(Note note) {
+                List<Index> indexes = new ArrayList<Index>(1);
+                indexes.add(new Index("null_column", null));
                 return indexes;
             }
         };

--- a/Simperium/src/main/java/com/simperium/android/PersistentStore.java
+++ b/Simperium/src/main/java/com/simperium/android/PersistentStore.java
@@ -191,10 +191,10 @@ public class PersistentStore implements StorageProvider {
                     values.put(key, (Long) value);
                 } else if(value instanceof Boolean){
                     values.put(key, (Boolean) value);
-                } else {
+                } else if(value != null) {
                     values.put(key, value.toString());
                 }
-                database.insertOrThrow(INDEXES_TABLE, null, values);
+                database.insertOrThrow(INDEXES_TABLE, key, values);
             }
 
             // If we have a fulltext index, let's add a record

--- a/Simperium/src/main/java/com/simperium/android/PersistentStore.java
+++ b/Simperium/src/main/java/com/simperium/android/PersistentStore.java
@@ -487,17 +487,19 @@ public class PersistentStore implements StorageProvider {
                 names.add(condition.getKey());
                 filters.append(String.format(Locale.US, " LEFT JOIN indexes AS i%d ON objects.bucket = i%d.bucket AND objects.key = i%d.key AND i%d.name=?", i, i, i, i));
                 Object subject = condition.getSubject();
-                String null_condition = condition.includesNull() ? String.format(Locale.US, " i%d.value IS NULL OR", i) : String.format(Locale.US, " i%d.value IS NOT NULL AND", i);
-                where.append(String.format(Locale.US, " AND ( %s i%d.value %s ", null_condition, i, condition.getComparisonType()));
-                if (subject instanceof Float) {
-                    where.append(String.format(Locale.US, " %f)", (Float)subject));
-                } else if (subject instanceof Integer){
-                    where.append(String.format(Locale.US, " %d)", (Integer)subject));
-                } else if (subject instanceof Boolean){
-                    where.append(String.format(Locale.US, " %d)", ((Boolean)subject ? 1 : 0)));
-                } else if (subject != null) {
-                    where.append(" ?)");
-                    replacements.add(subject.toString());
+                if (subject != null) {
+                    String null_condition = condition.includesNull() ? String.format(Locale.US, " i%d.value IS NULL OR", i) : String.format(Locale.US, " i%d.value IS NOT NULL AND", i);
+                    where.append(String.format(Locale.US, " AND ( %s i%d.value %s ", null_condition, i, condition.getComparisonType()));
+                    if (subject instanceof Float) {
+                        where.append(String.format(Locale.US, " %f)", (Float)subject));
+                    } else if (subject instanceof Integer){
+                        where.append(String.format(Locale.US, " %d)", (Integer)subject));
+                    } else if (subject instanceof Boolean){
+                        where.append(String.format(Locale.US, " %d)", ((Boolean)subject ? 1 : 0)));
+                    } else if (subject != null) {
+                        where.append(" ?)");
+                        replacements.add(subject.toString());
+                    }
                 }
                 i++;
             }

--- a/Simperium/src/main/java/com/simperium/android/PersistentStore.java
+++ b/Simperium/src/main/java/com/simperium/android/PersistentStore.java
@@ -487,20 +487,46 @@ public class PersistentStore implements StorageProvider {
                 names.add(condition.getKey());
                 filters.append(String.format(Locale.US, " LEFT JOIN indexes AS i%d ON objects.bucket = i%d.bucket AND objects.key = i%d.key AND i%d.name=?", i, i, i, i));
                 Object subject = condition.getSubject();
-                if (subject != null) {
-                    String null_condition = condition.includesNull() ? String.format(Locale.US, " i%d.value IS NULL OR", i) : String.format(Locale.US, " i%d.value IS NOT NULL AND", i);
-                    where.append(String.format(Locale.US, " AND ( %s i%d.value %s ", null_condition, i, condition.getComparisonType()));
-                    if (subject instanceof Float) {
-                        where.append(String.format(Locale.US, " %f)", (Float)subject));
-                    } else if (subject instanceof Integer){
-                        where.append(String.format(Locale.US, " %d)", (Integer)subject));
-                    } else if (subject instanceof Boolean){
-                        where.append(String.format(Locale.US, " %d)", ((Boolean)subject ? 1 : 0)));
-                    } else {
-                        where.append(" ?)");
-                        replacements.add(subject.toString());
+
+                // short circuit for null subjects
+                if (subject == null) {
+
+                    switch(condition.getComparisonType()) {
+
+                        case EQUAL_TO :
+                        case LIKE :
+                            where.append(String.format(Locale.US, " AND ( i%d.value IS NULL ) ", i));
+                            break;
+
+                        case NOT_EQUAL_TO :
+                        case NOT_LIKE :
+                            where.append(String.format(Locale.US, " AND ( i%d.value NOT NULL ) ", i));
+                            break;
+
+                        default :
+                            // noop
+                            break;
+
                     }
+
+                    i++;
+
+                    continue;
                 }
+
+                String null_condition = condition.includesNull() ? String.format(Locale.US, " i%d.value IS NULL OR", i) : String.format(Locale.US, " i%d.value IS NOT NULL AND", i);
+                where.append(String.format(Locale.US, " AND ( %s i%d.value %s ", null_condition, i, condition.getComparisonType()));
+                if (subject instanceof Float) {
+                    where.append(String.format(Locale.US, " %f)", (Float)subject));
+                } else if (subject instanceof Integer){
+                    where.append(String.format(Locale.US, " %d)", (Integer)subject));
+                } else if (subject instanceof Boolean){
+                    where.append(String.format(Locale.US, " %d)", ((Boolean)subject ? 1 : 0)));
+                } else if (subject != null) {
+                    where.append(" ?)");
+                    replacements.add(subject.toString());
+                }
+
                 i++;
             }
 

--- a/Simperium/src/main/java/com/simperium/android/PersistentStore.java
+++ b/Simperium/src/main/java/com/simperium/android/PersistentStore.java
@@ -496,7 +496,7 @@ public class PersistentStore implements StorageProvider {
                         where.append(String.format(Locale.US, " %d)", (Integer)subject));
                     } else if (subject instanceof Boolean){
                         where.append(String.format(Locale.US, " %d)", ((Boolean)subject ? 1 : 0)));
-                    } else if (subject != null) {
+                    } else {
                         where.append(" ?)");
                         replacements.add(subject.toString());
                     }

--- a/Simperium/src/main/java/com/simperium/android/PersistentStore.java
+++ b/Simperium/src/main/java/com/simperium/android/PersistentStore.java
@@ -47,7 +47,7 @@ public class PersistentStore implements StorageProvider {
         return new DataStore<T>(bucketName, schema);
     }
 
-    private class DataStore<T extends Syncable> implements BucketStore<T> {
+    protected class DataStore<T extends Syncable> implements BucketStore<T> {
 
         protected BucketSchema<T> schema;
         protected String bucketName;
@@ -409,16 +409,16 @@ public class PersistentStore implements StorageProvider {
         return database.rawQuery(String.format(Locale.US, "PRAGMA table_info(`%s`)", tableName), null);
     }
 
-    private class QueryBuilder {
+    protected static class QueryBuilder {
 
         private Query query;
-        private StringBuilder selection;
-        private String statement;
-        private String[] args;
         private DataStore mDataStore;
+        protected StringBuilder selection;
+        protected String statement;
+        protected String[] args;
 
         QueryBuilder(DataStore store, Query query){
-            this.mDataStore = store;
+            mDataStore = store;
             this.query = query;
             compileQuery();
         }

